numpy/core/src/umath/loops_minmax.dispatch.c.src

```diff
@@ -0,0 +1,671 @@
+/*@targets
+ ** $maxopt baseline
+ ** neon
+ **/
+#define _UMATHMODULE
+#define _MULTIARRAYMODULE
+#define NPY_NO_DEPRECATED_API NPY_API_VERSION
+
+#include "simd/simd.h"
+#include "loops_utils.h"
+#include "loops.h"
+#include "lowlevel_strided_loops.h"
+// Provides the various *_LOOP macros
+#include "fast_loop_macros.h"
+
```

Comment:
the whole template repeats need to be redesigned in order to fit all SIMD extensions, here an ideal design of repeats can fit any kind of SIMD extensions and friendly with universal intrinsics:

<details>
<summary>Expand me</summary>

```c

/********************************************************************************
 ** Scalar intrinsics
 ********************************************************************************/
// signed/unsigned int
#define c_max_i(A, B) ((A > B) ? A : B)
#define c_min_i(A, B) ((A < B) ? A : B)
// fp, propagates NaNs
#define c_max_f(A, B) ((A >= B || npy_isnan(A)) ? A : B)
#define c_max_d c_max_f
#define c_max_l c_max_f
#define c_min_f(A, B) ((A <= B || npy_isnan(A)) ? A : B)
#define c_min_d c_min_f
#define c_min_l c_min_f
// fp, ignores NaNs
#define c_maxp_f fmaxf
#define c_maxp_d fmax
#define c_maxp_l fmaxl
#define c_minp_f fminf
#define c_minp_d fmin
#define c_minp_l fminl

/********************************************************************************
 ** extra SIMD intrinsics
 ********************************************************************************/
#if NPY_SIMD
/**begin repeat
 * #sfx = s8, u8, s16, u16, s32, u32, s64, u64#
 */
#if defined(NPY_HAVE_ASIMD) && defined(__aarch64__)
    // arm64
#else
    // TODO eumlate by aligned store -> loop scalars
    NPY_FINLINE npyv_lanetype_@sfx@ npyv_reduce_min_@sfx@(npyv_@sfx@ a)
    {
    }
    NPY_FINLINE npyv_lanetype_@sfx@ npyv_reduce_max_@sfx@(npyv_@sfx@ a)
    {
    }
#endif
/**end repeat**/
#endif // NPY_SIMD

/**begin repeat
 * #sfx = f32, f64#
 * #bsfx = b32, b64#
 * #simd_chk = NPY_SIMD, NPY_SIMD_F64#
 * #scalar_sfx = f, d#
 */
#if @simd_chk@
#if defined(NPY_HAVE_ASIMD) && defined(__aarch64__)
    // arm64
#else
    // propagates NaNs
    NPY_FINLINE npyv_@sfx@ npyv_minn_@sfx@(npyv_@sfx@ a, npyv_@sfx@ b)
    {
        // TODO: use notnan, min, select
    }
    NPY_FINLINE npyv_@sfx@ npyv_maxn_@sfx@(npyv_@sfx@ a, npyv_@sfx@ b)
    {
        // TODO: use notnan, max, select
    }
    // TODO eumlate the following by aligned store -> loop scalars
    /**begin repeat1
     * #intrin = minn, maxn, minp, maxp#
     */
    NPY_FINLINE npyv_lanetype_@sfx@ npyv_reduce_@intrin@_@sfx@(npyv_@sfx@ a)
    {
    }
    /**end repeat1**/
#endif
#endif // simd_chk
/**end repeat**/

/********************************************************************************
 ** Defining the SIMD kernels
 ********************************************************************************/

/**begin repeat
 * #sfx = s8, u8, s16, u16, s32, u32, s64, u64, f32, f64#
 * #simd_chk = NPY_SIMD*9, NPY_SIMD_F64#
 * #is_fp = 0*8, 1, 1#
 * #scalar_sfx = i*8, f, d#
 */
/**begin repeat1
 * # intrin = max, min, maxp, minp#
 * # fp_only = 0, 0, 1, 1#
 */
#if @simd_chk@ && (!@fp_only@ || (@is_fp@ && @fp_only@))
#undef V_INTRIN
#undef V_REDUCE_INTRIN
#if @is_fp@ && !@fp_only@
    #define V_INTRIN npyv_@intrin@n_@sfx@ // propagates NaNs
    #define V_REDUCE_INTRIN npyv_reduce_@intrin@n_@sfx@
#else
    #define V_INTRIN npyv_@intrin@_@sfx@
    #define V_REDUCE_INTRIN npyv_reduce_@intrin@_@sfx@
#endif
// contiguous input.
static inline void
simd_reduce_c_@intrin@_@sfx@(const npyv_lanetype_@sfx@ *ip, npyv_lanetype_@sfx@ *op1, npy_intp len)
{
}
// contiguous inputs and output.
static inline void
simd_binary_ccc_@intrin@_@sfx@(const npyv_lanetype_@sfx@ *ip1, const npyv_lanetype_@sfx@ *ip2,
                                     npyv_lanetype_@sfx@ *op1, npy_intp len)
{
    // TODO: simd binary op goes here
}
#endif // simd_chk && (!fp_only || (is_fp && fp_only))
/**end repeat1**/
/**end repeat**/

/********************************************************************************
 ** Defining ufunc inner functions
 ********************************************************************************/
/**begin repeat
 * #TYPE = ULONG, UBYTE, USHORT, UINT, ULONGLONG,
 *         BYTE, SHORT, INT, LONG, LONGLONG,
 *         FLOAT, DOUBLE, LONGDOUBLE#
 *
 * #BTYPE = LONG, BYTE, SHORT, UINT, LONGLONG,
 *          BYTE, SHORT, INT, LONG, LONGLONG,
 *          FLOAT, DOUBLE, LONGDOUBLE#
 * #type = npy_ulong, npy_ubyte, npy_ushort, npy_uint, npy_ulonglong,
 *         npy_byte, npy_short, npy_int, npy_long, npy_longlong,
 *         float, double, long double#
 *
 * #is_fp = 0*10, 1*3#
 * #is_unsigend = 1*5, 0*5, 0*3#
 * #scalar_sfx = i*10, f, d, l#
 */
#undef TO_SIMD_SFX
#if 0
/**begin repeat1
 * #len = 8, 16, 32, 64#
 */
#elif NPY_SIMD && NPY_BITSOF_@BTYPE@ == @len@
    #if @is_fp@
        #define TO_SIMD_SFX(X) X##_f@len@
        #if NPY_BITSOF_@BTYPE@ == 64 && !NPY_SIMD_F64
            #undef TO_SIMD_SFX
        #endif
    #elif @is_unsigend@
        #define TO_SIMD_SFX(X) X##_u@len@
    #else
        #define TO_SIMD_SFX(X) X##_s@len@
    #endif
/**end repeat1**/
#endif

/**begin repeat1
 * # kind = maximum, minimum, fmax, fmin#
 * # intrin = max, min, maxp, minp#
 * # fp_only = 0, 0, 1, 1#
 */
#if !@fp_only@ || (@is_fp@ && @fp_only@)
NPY_NO_EXPORT void NPY_CPU_DISPATCH_CURFX(@TYPE@_@kind@)
(char **args, npy_intp const *dimensions, npy_intp const *steps, void *NPY_UNUSED(func))
{
    char *ip1 = args[0], *ip2 = args[1], *op1 = args[2];
    npy_intp is1 = steps[0], is2 = steps[1], os1 = steps[2],
             len = dimensions[0];
#ifdef TO_SIMD_SFX
    #undef STYPE
    #define STYPE TO_SIMD_SFX(npyv_lanetype)
    if (IS_BINARY_REDUCE) {
        // reduce and contiguous
        if (is2 == sizeof(@type@)) {
            TO_SIMD_SFX(simd_reduce_c_@intrin@)(
                (STYPE*)ip2, (STYPE*)op1, len
            );
            goto clear_fp;
        }
    }
    if (!is_mem_overlap(ip1, is1, op1, os1, len) &&
        !is_mem_overlap(ip2, is2, op1, os1, len)
    ) {
        // no overlap and operands are binary contiguous
        if (IS_BINARY_CONT(@type@, @type@)) {
            TO_SIMD_SFX(simd_binary_ccc_@intrin@)(
                (STYPE*)ip1, (STYPE*)ip2, (STYPE*)op1, len
            );
            goto clear_fp;
        }
    }
#endif // TO_SIMD_SFX
#ifndef NPY_DISABLE_OPTIMIZATION
    // any scalar unroll goes here
#endif // NPY_DISABLE_OPTIMIZATION
    for (; len > 0; --len, ip1 += is1, ip2 += is2, op1 += os1) {
        const @type@ in1 = *(@type@ *)ip1;
        const @type@ in2 = *(@type@ *)ip2;
        *((@type@ *)op1) = c_@intrin@_@scalar_sfx@(in1, in2);
    }
#ifdef TO_SIMD_SFX
clear_fp:;
#endif
#if @is_fp@
    npy_clear_floatstatus_barrier((char*)dimensions);
#endif
}
#endif // !fp_only || (is_fp && fp_only)
/**end repeat1**/
/**end repeat**/

```
</details>